Update wash build with build_command (#292)
wasmCloud/wasmCloud.com: docs/cli/build.md

description: "wash build command reference"
---

> `wash build` enables users to build and sign custom wasmCloud entities like actors, providers, or interfaces. One can bring in their own project containing some business logic and build these entities by simply providing the path to the project or its associated `wasmcloud.toml` file. The built artifact is signed using automatically generated keys or a user may use their own keys to sign the build.

Reviewer: ```suggestion
`wash build` enables you to build and sign custom wasmCloud entities like actors, providers, or interfaces. You can bring in your own project containing some business logic and build these entities by simply providing the path to the project or its associated `wasmcloud.toml` file. The built artifact is signed using automatically generated keys or you may use your own keys to sign the build.
```

### Usage

```
wash build -p path/to/wasmcloud.toml
```

#### Custom build commands

In `wasmcloud.toml`, the `build_command` field can specify a custom build command to run "under the hood" of `wash build`. This enables extended language support through any component toolchain that compiles to WASI 0.2. The example below uses [`componentize-py`](https://github.com/bytecodealliance/componentize-py) to generate a Python-based component:

```toml
name = "PythonExample"
language = "rust"
type = "actor"
version = "0.1.0"

[actor]
claims = ["wasmcloud:httpclient", "wasmcloud:httpserver"]
build_command = "componentize-py -d ../../wit -w wasi:http/proxy@0.2.0-rc-2023-12-05 componentize app -o http.wasm"
build_artifact = "http.wasm"
```

> When using the `build_command` field, users must also specify the `build_artifact`: the filename and location of the Wasm artifact that will be produced by the build command. This provides `wash build` with a target for signing the artifact.  

Reviewer: ```suggestion
If you use the `build_command` field, you must also specify the `build_artifact`: the filename and location of the Wasm artifact that will be produced by the build command. This provides `wash build` with a target for signing the artifact.
```

> **Note**: Currently the `language` field in the `wasmcloud.toml` metadata must be set to an "officially supported" language (i.e. Rust or Go) to avoid errors. The `build_command` field does not have full support for environment variables or multiple commands and should be in the form of "command arg1 arg2 arg...". Use an external script to handle more complex build commands.  

Reviewer 2: NIT: we do have a callout syntax (`::info`? something like that), but can be fixed in a follow up!

Reviewer: @vados-cosmonic, good point.

@ericgregory These are called [admonitions](https://docusaurus.io/docs/markdown-features/admonitions) and they work like this:
https://github.com/wasmCloud/wasmcloud.com/blob/d6e9e1bfb3932e829293e817d19a6a64f84638ed/docs/ecosystem/wadm/index.md?plain=1#L16-L18

Author: @vados-cosmonic @lachieh Ah, that syntax will definitely be helpful—thanks both!
